botorch/acquisition/knowledge_gradient.py

```diff
@@ -46,12 +47,456 @@
+
+        # generate equal quantile spaced z_vals
+        z_vals = self.construct_z_vals(self.num_fantasies)
+
+        kgvals = torch.zeros(X.shape[0], dtype=torch.double)
+        for xnew_idx, xnew in enumerate(X):
+            # Compute X discretisation using the different generated quantiles.
+            x_star, _ = self.compute_mc_kg(xnew, z_vals)
+
+            # Compute value of discrete Knowledge Gradient using the generated discretisation
+            kgvals[xnew_idx] = self.compute_discrete_kg(
+                xnew=xnew, optimal_discretisation=x_star
+            )
+
+        return kgvals
+
+    def compute_discrete_kg(
+        self, xnew: Tensor, optimal_discretisation: Tensor
+    ) -> Tensor:
+        """
+
+        Args:
+        xnew: A `1 x 1 x d` Tensor with `1` acquisition function evaluations of
+            `d` dimensions.
+            optimal_discretisation: num_fantasies x d Tensor. Optimal X values for each z in Zvals.
+
+        """
+        # Augment the discretisation with the designs.
+        concatenated_xnew_discretisation = torch.cat(
+            [xnew, optimal_discretisation], dim=0
+        ).squeeze()  # (m + num_X_disc, d)
+
+        # Compute posterior mean, variance, and covariance.
+        full_posterior = self.model.posterior(
+            concatenated_xnew_discretisation, observation_noise=False
+        )
+        noise_variance = self.model.likelihood.noise_covar.noise
+        full_posterior_mean = full_posterior.mean  # (1 + num_X_disc , 1)
+
+        # Compute full Covariante Cov(Xnew, X_discretised), select [Xnew X_discretised] submatrix, and subvectors.
+        discretisation_mean = full_posterior_mean[
+            len(xnew) :
+        ].squeeze()  # (num_X_disc, )
+        full_posterior_covariance = (
+            full_posterior.mvn.covariance_matrix
+        )  # (1 + num_X_disc , 1 + num_X_disc )
+        posterior_cov_xnew_opt_disc = full_posterior_covariance[
+            : len(xnew), :
+        ].squeeze()  # ( 1 + num_X_disc,)
+        full_posterior_variance = (
+            full_posterior.variance.squeeze()
+        )  # (1 + num_X_disc, )
+        full_predictive_covariance = (
+            posterior_cov_xnew_opt_disc
+            / (full_posterior_variance + noise_variance).sqrt()
+        )
+        # initialise empty kgvals torch.tensor
+        kgval = self.kgcb(a=full_posterior_mean, b=full_predictive_covariance)
+
+        return kgval
+
+    def compute_mc_kg(self, xnew: Tensor, Zvals: Tensor) -> tuple[Tensor, Tensor]:
+        """
+
+        Args:
+            xnew: A `1 x 1 x d` Tensor with `1` acquisition function evaluations of
+            `d` dimensions.
+            Zvals: 1 x num_fantasies Tensor with num_fantasies Normal quantiles.
+
+        Returns:
+            xstar_inner_optimisation: num_fantasies x d Tensor. Optimal X values for each z in Zvals.
+            xnew_acquisition_value: 1 x 1 Tensor, Monte Carlo KG value of xnew
```

Comment:
the second output variable is not in the X space, rename it? ~`xnew_acquisition_value`~ -> `kg_estimate`?

Reply:
done